qiskit/dagcircuit/dagdependency_v2.py

```diff
@@ -0,0 +1,740 @@
+            DAGDependencyError: If the supplied :obj:`~Bit` could not be found on the circuit.
+        """
+        try:
+            if isinstance(bit, Qubit):
+                return self._qubit_indices[bit]
+            elif isinstance(bit, Clbit):
+                return self._clbit_indices[bit]
+            else:
+                raise DAGDependencyError(f"Could not locate bit of unknown type: {type(bit)}")
+        except KeyError as err:
+            raise DAGDependencyError(
+                f"Could not locate provided bit: {bit}. Has it been added to the DAGDependency?"
+            ) from err
+
+    def apply_operation_back(self, operation, qargs=(), cargs=()):
+        """Add a DAGOpNode to the graph and update the edges.
+
+        Args:
+            operation (qiskit.circuit.Operation): operation as a quantum gate
+            qargs (list[~qiskit.circuit.Qubit]): list of qubits on which the operation acts
+            cargs (list[Clbit]): list of classical wires to attach to
+        """
+        new_node = DAGOpNode(
+            op=operation,
+            qargs=qargs,
+            cargs=cargs,
+            dag=self,
+        )
+        new_node._node_id = self._multi_graph.add_node(new_node)
+        self._update_edges()
+        self._increment_op(new_node.op)
+
+        directives = ["measure"]
+        if not getattr(operation, "_directive", False) and operation.name not in directives:
+            self.qindices_map[new_node] = [self.find_bit(q).index for q in qargs]
+            if getattr(operation, "condition", None):
+                # The change to handling operation.condition follows code patterns in quantum_circuit.py.
+                # However:
+                #   (1) cindices_map is specific to template optimization and should not be computed
+                #       in this place.
+                #   (2) Template optimization pass currently does not handle general conditions.
+                cond_bits = condition_resources(operation.condition).clbits
+                self.cindices_map[new_node] = [self.find_bit(c).index for c in cond_bits]
+            else:
+                self.cindices_map[new_node] = []
+        else:
+            self.qindices_map[new_node] = []
+            self.cindices_map[new_node] = []
+
+    def _update_edges(self):
+        """
+        Updates DagDependencyV2 by adding edges to the newly added node (max_node)
+        from the previously added nodes.
+        For each previously added node (prev_node), an edge from prev_node to max_node
+        is added if max_node is "reachable" from prev_node (this means that the two
+        nodes can be made adjacent by commuting them with other nodes), but the two nodes
+        themselves do not commute.
+
+        Currently. this function is only used when creating a new DAGDependencyV2 from another
+        representation of a circuit, and hence there are no removed nodes (this is why
+        iterating over all nodes is fine).
+        """
```

Comment:
This is just a warning: this function internally assumes that the newly added node has the largest node index and is added at the end of the graph. As we are extending the functionality of DAG dependency (and in particular adding methods to remove nodes), I am afraid that something here might break. Possibly it would be best to eventually rewrite this function to be more general.

Reply:
As I mentioned above, this is one I took a first try at rewriting to use node object ids. I agree it should be redone to avoid iterating over the indices.